author	Anderson Lizardo <anderson.lizardo@openbossa.org>
commit	371e9dfbe94343154146e7742abb80c30c3c2ad4
	Wed, 4 Apr 2012 10:00:03 +0000 (13:00 +0300)

thermometer: Fix D-Bus service disconnect watch removal

destroy_watcher() will be automatically called when removing the D-Bus
watch (it was set as the destroy callback when adding the watch),
therefore it is only necessary to remove the watch.

diff -cr bluez-4.99.pan.orig/thermometer/thermometer.c bluez-4.99.pan/thermometer/thermometer.c
*** bluez-4.99.pan.orig/thermometer/thermometer.c	Mon Mar  5 04:03:57 2012
--- bluez-4.99.pan/thermometer/thermometer.c	Mon Oct 15 03:09:30 2012
***************
*** 136,149 ****
  {
  	struct watcher *watcher = user_data;
  
- 	if (watcher->id > 0)
- 		g_dbus_remove_watch(watcher->t->conn, watcher->id);
- 
  	g_free(watcher->path);
  	g_free(watcher->srv);
  	g_free(watcher);
  }
  
  static void destroy_char(gpointer user_data)
  {
  	struct characteristic *c = user_data;
--- 136,153 ----
  {
  	struct watcher *watcher = user_data;
  
  	g_free(watcher->path);
  	g_free(watcher->srv);
  	g_free(watcher);
  }
  
+ static void remove_watcher(gpointer user_data)
+ {
+ 	struct watcher *watcher = user_data;
+ 
+ 	g_dbus_remove_watch(watcher->t->conn, watcher->id);
+ }
+ 
  static void destroy_char(gpointer user_data)
  {
  	struct characteristic *c = user_data;
***************
*** 172,178 ****
  		g_slist_free_full(t->chars, destroy_char);
  
  	if (t->fwatchers != NULL)
! 		g_slist_free_full(t->fwatchers, destroy_watcher);
  
  	dbus_connection_unref(t->conn);
  	btd_device_unref(t->dev);
--- 176,182 ----
  		g_slist_free_full(t->chars, destroy_char);
  
  	if (t->fwatchers != NULL)
! 		g_slist_free_full(t->fwatchers, remove_watcher);
  
  	dbus_connection_unref(t->conn);
  	btd_device_unref(t->dev);
***************
*** 814,820 ****
  	remove_int_watcher(t, watcher);
  
  	t->fwatchers = g_slist_remove(t->fwatchers, watcher);
! 	watcher->id = 0;
  
  	if (g_slist_length(t->fwatchers) == 0)
  		disable_final_measurement(t);
--- 818,824 ----
  	remove_int_watcher(t, watcher);
  
  	t->fwatchers = g_slist_remove(t->fwatchers, watcher);
! 	g_dbus_remove_watch(watcher->t->conn, watcher->id);
  
  	if (g_slist_length(t->fwatchers) == 0)
  		disable_final_measurement(t);
***************
*** 893,899 ****
  	remove_int_watcher(t, watcher);
  
  	t->fwatchers = g_slist_remove(t->fwatchers, watcher);
! 	destroy_watcher(watcher);
  
  	if (g_slist_length(t->fwatchers) == 0)
  		disable_final_measurement(t);
--- 897,903 ----
  	remove_int_watcher(t, watcher);
  
  	t->fwatchers = g_slist_remove(t->fwatchers, watcher);
! 	g_dbus_remove_watch(watcher->t->conn, watcher->id);
  
  	if (g_slist_length(t->fwatchers) == 0)
  		disable_final_measurement(t);
